src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,746 @@
+            let e = box_err!("snapshot of lock cf applied from sst file");
+            error!("cannot collect all applied lock"; "err" => ?e);
+            self.send(LockObserverMsg::Err(e));
+        }
+    }
+}
+
+struct LockCollectorRunner {
+    observer_state: Arc<LockObserverState>,
+
+    collected_locks: Vec<(Key, Lock)>,
+
+    /// `is_clean` is true, only it's sure that all applying of stale locks (locks with start_ts <=
+    /// specified max_ts) are monitored and collected. If there are too many stale locks or any
+    /// error happens, `is_clean` must be set to `false`.
+    /// TODO: when dirty, stop collecting.
+    is_clean: bool,
+}
+
+impl LockCollectorRunner {
+    pub fn new(observer_state: Arc<LockObserverState>) -> Self {
+        Self {
+            observer_state,
+            collected_locks: vec![],
+            is_clean: true,
+        }
+    }
+
+    fn handle_observer_msg(&mut self, msg: LockObserverMsg) {
+        if !self.collected_locks.len() >= MAX_COLLECT_SIZE {
+            return;
+        }
+
+        match msg {
+            LockObserverMsg::Err(e) => {
+                if self.is_clean {
+                    self.is_clean = false;
+                    info!("lock collector marked dirty because received error"; "err" => ?e);
+                }
+            }
+            LockObserverMsg::Locks(mut locks) => {
+                if locks.len() + self.collected_locks.len() > MAX_COLLECT_SIZE {
+                    self.is_clean = false;
+                    info!("lock collector marked dirty because received too many locks");
+                    locks.truncate(MAX_COLLECT_SIZE - self.collected_locks.len());
+                }
+                self.collected_locks.extend(locks);
+            }
+        }
+    }
+
+    fn start_collecting(&mut self, max_ts: TimeStamp) -> Result<()> {
+        if self.observer_state.max_ts.load(Ordering::Acquire) == max_ts {
+            // Repeated request. Ignore it.
+            return;
+        }
+        info!("start collecting locks"; "max_ts" => max_ts);
+        self.collected_locks.clear();
+        self.is_clean = true;
+        self.observer_state
+            .max_ts
+            .store(max_ts.into_inner(), Ordering::Release);
+        Ok(())
+    }
+
+    fn get_collected_locks(&mut self, max_ts: TimeStamp) -> Result<(Vec<LockInfo>, bool)> {
+        let curr_max_ts = self.observer_state.max_ts.load(Ordering::Acquire);
+        let curr_max_ts = TimeStamp::new(curr_max_ts);
+        if curr_max_ts != max_ts {
+            warn!(
+                "trying to fetch collected locks but now collecting with another max_ts";
+                "req_max_ts" => max_ts,
+                "current_max_ts" => curr_max_ts,
+            );
+            return Err(box_err!(
+                "trying to fetch collected locks but now collecting with another max_ts"
+            ));
+        }
+
+        let locks: Result<_> = self
+            .collected_locks
```

Comment:
How about using `mem:::replace` to avoid clone?

Reply:
There can be several get requests. I think they should be idempotent.

Reply:
Oh, yes.

Reply:
If we drop them here, then if the response is lost in the network, we will never get the result then.